File: 1 - metodosBusca/metodosBusca/src/Hannoi.java
import java.util.LinkedList;
import java.util.List;

import busca.BuscaLargura;
import busca.BuscaProfundidade;
import busca.Estado;
import busca.MostraStatusConsole;
import busca.Nodo;
import java.util.HashSet;
import java.util.Stack;
import javax.swing.JOptionPane;

public class Hannoi implements Estado {
    
    @Override
    public String getDescricao() {
        return " Problema das torres de Hannoi.... ";
    }
    final Stack torre1, torre2, torre3;
    final String op; // operacao que gerou o estado
    
    /** construtor para o estado, recebe cada valor de atributo */
    public Hannoi(Stack torre1, Stack torre2, Stack torre3, String op) {
        this.torre1 = torre1;
        this.torre2 = torre2;
        this.torre3 = torre3;
        
        this.op = op;
    }
    
    /**
     * verifica se o estado final foi atingido
     */
    @Override
    public boolean ehMeta() {
        return this.torre1.isEmpty() && this.torre2.isEmpty();
    }
    
    /**
     * Custo para geracao do estado
     */
    @Override
    public int custo() {
        return 1;
    }
    
    /**
     * gera uma lista de sucessores do nodo.
     * @return visitados
     */
    @Override
    public List<Estado> sucessores() {
        List<Estado> visitados = new LinkedList<Estado>(); // a lista de sucessores
        HashSet<Estado> visitadosTemp = new HashSet<>();
        
        moverTorre1_Torre2(visitadosTemp);
        moverTorre1_Torre3(visitadosTemp);
        moverTorre2_Torre1(visitadosTemp);
        moverTorre2_Torre3(visitadosTemp);
        moverTorre3_Torre1(visitadosTemp);
        moverTorre3_Torre2(visitadosTemp);
        
        visitados.clear();
        visitados.addAll(visitadosTemp);
        
        return visitados;
    }
    
    private boolean ehValido(Stack origem, Stack destino) {
        if (origem.isEmpty() ||
            (!origem.isEmpty() && !destino.isEmpty() && (int)origem.peek() > (int)destino.peek())) {
            return false;
        }
        
        return true;
    }
    
    private void moverTorre1_Torre2(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre1.clone();
        Stack destino = (Stack)this.torre2.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi(origem, destino, (Stack)this.torre3.clone(), "Movendo torre1 para torre2");
            visitados.add(novo);
        }
    }
    
    private void moverTorre1_Torre3(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre1.clone();
        Stack destino = (Stack)this.torre3.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi(origem, (Stack)this.torre2.clone(), destino, "Movendo torre1 para torre3");
            visitados.add(novo);
        }
    }
    
    private void moverTorre2_Torre1(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre2.clone();
        Stack destino = (Stack)this.torre1.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi(destino, origem, (Stack)this.torre3.clone(), "Movendo torre2 para torre1");
            visitados.add(novo);
        }
    }
    
    private void moverTorre2_Torre3(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre2.clone();
        Stack destino = (Stack)this.torre3.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi((Stack)this.torre1.clone(), origem, destino, "Movendo torre2 para torre3");
            visitados.add(novo);
        }
    }
    
    private void moverTorre3_Torre1(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre3.clone();
        Stack destino = (Stack)this.torre1.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi(destino, (Stack)this.torre2.clone(), origem, "Movendo torre3 para torre1");
            visitados.add(novo);
        }
    }
    
    private void moverTorre3_Torre2(HashSet<Estado> visitados) {
        Stack origem = (Stack)this.torre3.clone();
        Stack destino = (Stack)this.torre2.clone();
        
        if (ehValido(origem, destino)){
            destino.push(origem.pop());
            Hannoi novo = new Hannoi((Stack)this.torre1.clone(), destino, origem, "Movendo torre3 para torre2");
            visitados.add(novo);
        }
    }
    /**
     * verifica se um estado e igual a outro
     * (usado para poda)
     */
    @Override
    public boolean equals(Object o) {
        if (o instanceof Hannoi) {
            Hannoi e = (Hannoi)o;
            return this.torre1.equals(e.torre1) &&
                   this.torre2.equals(e.torre2) &&
                   this.torre3.equals(e.torre3);
        }
        return false;
    }
    
    /** 
     * retorna o hashCode do estado
     * (usado para poda, conjunto de fechados)
     */
    @Override
    public int hashCode() {
        return (""+this.torre1 + this.torre2 + this.torre3).hashCode();
    }
    
    @Override
    public String toString() {
        return "" + this.torre1 + this.torre2 + this.torre3  + " -> " + op + "\n";
    }
    
    
    public static void main(String[] a) {
        int discos = Integer.parseInt(JOptionPane.showInputDialog(null, "Quantos discos?"));
        if (discos == 0) System.exit(0);
        
        Stack t1 = new Stack();
        Stack t2 = new Stack();
        Stack t3 = new Stack();
        
        for (;discos > 0; discos--) {
            t1.push(discos);
        }
        
        Hannoi estadoInicial = new Hannoi(t1, t2, t3, "estado inicial");
        
        // chama busca em largura
        System.out.println("busca em ....");
        
        Nodo n = new BuscaProfundidade(new MostraStatusConsole()).busca(estadoInicial);
        
        if (n == null) {
            System.out.println("sem solucao!");
        } else {
            System.out.println("solucao:\n" + n.montaCaminho() + "\n\n");
        }
    }
}

File: 1 - metodosBusca/python/app/desafios/hannoi.py
from app.core.busca.busca_profundidade import BuscaProfundidade
from app.core.busca.estado import Estado
from app.core.busca.stack import Stack
from app.core.busca.mostra_status_console import MostraStatusConsole


class Hannoi(Estado):


    def __init__(self, torre1: Stack, torre2: Stack, torre3: Stack, op: str):
        super().__init__()
        self.torre1 = torre1
        self.torre2 = torre2
        self.torre3 = torre3
        self.op = op
    
    def get_descricao(self):
        return 'Problema das torres de Hannoi....'

    def sucessores(self) -> list:
        visitados: list[Estado] = []
        visitados_temp: list[Estado] = []
        self.mover_torre_1_torre_2(visitados=visitados_temp)
        self.mover_torre_1_torre_3(visitados=visitados_temp)
        self.mover_torre_2_torre_1(visitados=visitados_temp)
        self.mover_torre_2_torre_3(visitados=visitados_temp)
        self.mover_torre_3_torre_1(visitados=visitados_temp)
        self.mover_torre_3_torre_2(visitados=visitados_temp)

        visitados.clear()
        visitados.extend(visitados_temp)
        return visitados


    def __str__(self):
        return f'{self.torre1} {self.torre2} {self.torre3}->{self.op} \n'
    
    def __eq__(self, other):
        return self.torre1 == other.torre1 and self.torre2 == other.torre2 and self.torre3 == other.torre3

    def __hash__(self):
        return hash(f"{self.torre1}{self.torre2}{self.torre3}")

    def custo(self):
        return 1
    
    def eh_meta(self):
        return self.torre1.is_empty() and self.torre2.is_empty() and len(self.torre3) == 3
    
    def eh_valido(self, origem: Stack, destino: Stack) -> bool:
        if origem.is_empty():
            return False
        if destino.is_empty():
            return True
        return origem.peek() < destino.peek()
    
    def mover(self, origem: Stack, destino: Stack, aux: Stack, visitados: list[Estado]):
        if self.eh_valido(origem, destino):
            origem, destino, aux = origem.clone(), destino.clone(), aux.clone()
            destino.push(origem.pop())
            novo: Estado = Hannoi(origem, destino, aux, f'{origem} -> {destino}')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_1_torre_2(self, visitados: list[Estado]):
        origem, destino, aux = self.torre1.clone(), self.torre2.clone(), self.torre3.clone()
        if self.eh_valido(origem, destino):
            destino.push(origem.pop())
            novo: Estado = Hannoi(origem, destino, aux, 'Movendo torre 1 -> torre 2')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_1_torre_3(self, visitados: list[Estado]):
        origem, destino, aux = self.torre1.clone(), self.torre3.clone(), self.torre2.clone()
        if self.eh_valido(origem, destino):
            destino.push(origem.pop())
            novo: Estado = Hannoi(origem, aux, destino, 'Movendo torre 1 -> torre 3')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_2_torre_1(self, visitados: list[Estado]):
        origem, destino, aux = self.torre2.clone(), self.torre1.clone(), self.torre3.clone()
        if self.eh_valido(origem, destino):
            destino.push(origem.pop())
            novo: Estado = Hannoi(destino, origem, aux, 'Movendo torre 2 -> torre 1')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_2_torre_3(self, visitados: list[Estado]):
        origem, destino, aux = self.torre2.clone(), self.torre3.clone(), self.torre1.clone()
        if self.eh_valido(origem, destino):
            destino.push(origem.pop())
            novo: Estado = Hannoi(aux, origem, destino, 'Movendo torre 2 -> torre 3')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_3_torre_1(self, visitados: list[Estado]):
        origem, destino, aux = self.torre3.clone(), self.torre1.clone(), self.torre2.clone()
        if self.eh_valido(origem, destino):
            destino.push(origem.pop())
            novo: Estado = Hannoi(destino, aux, origem, 'Movendo torre 3 -> torre 1')
            
            if novo not in visitados:
                visitados.append(novo)
    
    def mover_torre_3_torre_2(self, visitados: list[Estado]):
        origem, destino, aux = self.torre3.clone(), self.torre2.clone(), self.torre1.clone()
        if self.eh_valido(origem, destino):
            origem, destino, aux = origem.clone(), destino.clone(), aux.clone()
            destino.push(origem.pop())
            novo: Estado = Hannoi(aux, destino, origem, 'Movendo torre 3 -> torre 2')
            
            if novo not in visitados:
                visitados.append(novo)


if __name__ == "__main__":
    discos = 3
    torre1 = Stack()
    torre2 = Stack()
    torre3 = Stack()
    for i in range(discos, 0, -1):
        torre1.push(i)
    estado = Hannoi(torre1, torre2, torre3, 'inicio')
    print(estado)

    console_show = MostraStatusConsole()

    node = BuscaProfundidade().busca(estado)
    if node is None:
        print('Não foi possivel encontrar uma solução')
    else:    
        while node is not None:
            print(node)
            node = node.pai
        print('Fim')
